Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (C) 2025 Apple Inc. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import subprocess
+import re
+from pathlib import Path
+from typing import Iterator, NamedTuple
+
+method_to_selector = re.compile(r'[-+]\[(?P<class>\S+) (?P<selector>[^\]]+)\]')
+
+
+class Report(NamedTuple):
+    exports: list[str]
+    selectors: set[str]
+    bindings: list[str]
+    selrefs: list[str]
+    file: Path
+
+
+def report(binary_path: Path, arch: str, symbols_only=False) -> Report:
+    # Symbol table
+    nm_lines = subprocess.check_output(
+        ('xcrun', 'nm', '-P', '--arch', arch, binary_path),
+        text=True, errors='replace'
+    ).splitlines()
+
+    exports = []
+    selectors = set()
+    bindings = []
+    selrefs = []
+
+    for line in nm_lines:
+        if not line or f'for architecture {arch}' in line:
+            # nm always
+            continue
+        symbol, flag, *_ = line.rsplit(maxsplit=4)
+        if flag in 'ATDBCS':
+            exports.append(symbol)
+        elif flag == 't':
+            m = method_to_selector.match(symbol)
+            if m:
+                selectors.add(m.group('selector'))
+        elif flag == 'U':
+            bindings.append(symbol)
+
+    if not symbols_only:
```

Comment:
Recommend some sort of early return or helper function call - that's a lot of indent.